Sources/ConcurrencyHelpers/Spinlock.swift

```diff
@@ -0,0 +1,59 @@
+// Copyright 2023 Ordo One AB
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+
+import Atomics
+import _PauseShims
+
+/// Lock to protect very short critical sections.
+public final class Spinlock {
+    private typealias State = Bool
+
+    private static let locked: State = true
+    private static let unlocked: State = false
+
+    private let stateStorage = UnsafeMutablePointer<State.AtomicRepresentation>.allocate(capacity: 1)
+    private var state: UnsafeAtomic<State>
+
+    /// Create a new spin-lock.
+    public init() {
+        stateStorage.initialize(to: State.AtomicRepresentation(Self.unlocked))
+        state = .init(at: stateStorage)
+    }
+
+    deinit {
+        stateStorage.deallocate()
+    }
+
+    /// Acquire the lock.
+    ///
+    /// Whenever possible, consider using `Lockable.withLock()` instead of this method and
+    /// `unlock`, to simplify lock handling.
+    public func lock() {
+        while state.load(ordering: .relaxed) != Self.unlocked {
+            _concurrency_helpers_pause()
+        }
+
+        while !state.compareExchange(expected: Self.unlocked, desired: Self.locked, ordering: .acquiringAndReleasing).exchanged {
```

Comment:
We wait while the sate is not locked with relaxed memory access to reduce number of bus locks...
But if we failed to lock then we just doing a busy wait.
Probably would be nice to wait while the state is locked with relaxed read as well.
I mean having an implementation like
```
    public func lock() {
        while !state.compareExchange(expected: Self.unlocked, desired: Self.locked, ordering: .acquiringAndReleasing).exchanged {
            while state.load(ordering: .relaxed) != Self.unlocked {
                _concurrency_helpers_pause()
            }
        }
    }
```
what do you think?

Reply:
Agree!